File: ios/Classes/MultiPayPlugin.swift
import Flutter
import UIKit
import AlipaySDK
import UPPaymentControlMini

@objc public class MultiPayPlugin: NSObject, FlutterPlugin {

    static let shared = MultiPayPlugin()

    private var alipayAppId: String?
    private var wechatAppId: String?
    private var wechatUniversalLink: String?
    private var unionPayTnMode: String?
    private var alipayScheme: String = "alipay"
    private var unionPayScheme: String = "uppay"
    private var pendingResult: FlutterResult?
    private var paymentChannel: String = ""
    private var unionPayPendingResult: FlutterResult?

    @objc public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "multi_pay_channel",
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(shared, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "pay":
            guard let args = call.arguments as? [String: Any],
                  let channel = args["channel"] as? String,
                  let orderInfo = args["orderInfo"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing required arguments", details: nil))
                return
            }
            handlePay(channel: channel, orderInfo: orderInfo, result: result)

        case "isAppInstalled":
            guard let args = call.arguments as? [String: Any],
                  let channel = args["channel"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing channel", details: nil))
                return
            }
            result(handleIsAppInstalled(channel: channel))

        case "initSdk":
            if let args = call.arguments as? [String: Any] {
                alipayAppId = args["alipayAppId"] as? String
                wechatAppId = args["wechatAppId"] as? String
                wechatUniversalLink = args["wechatUniversalLink"] as? String
                unionPayTnMode = args["unionPayTnMode"] as? String
                alipayScheme = args["alipayScheme"] as? String ?? "alipay"
                unionPayScheme = args["unionPayScheme"] as? String ?? "uppay"

                if let appId = wechatAppId {
                    let universalLink = wechatUniversalLink ?? "https://your-universal-link.com/"
                    WxPayBridge.registerApp(appId, universalLink: universalLink)
                }
            }
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func handlePay(channel: String, orderInfo: String, result: @escaping FlutterResult) {
        pendingResult = result
        paymentChannel = channel

        switch channel {
        case "alipay":
            payWithAlipay(orderInfo: orderInfo)
        case "wechat":
            payWithWechat(orderInfo: orderInfo)
        case "unionPay":
            payWithUnionPay(orderInfo: orderInfo)
        default:
            result(FlutterError(code: "UNKNOWN_CHANNEL", message: "Unknown channel: \(channel)", details: nil))
            pendingResult = nil
        }
    }

    private func payWithAlipay(orderInfo: String) {
        guard let topController = UIApplication.shared.topViewController() else {
            pendingResult?(FlutterError(code: "NO_VIEW_CONTROLLER", message: "No view controller", details: nil))
            pendingResult = nil
            return
        }

        AlipaySDK.defaultService().payOrder(orderInfo, fromScheme: alipayScheme) { [weak self] resDict in
            guard let self = self else { return }
            self.didReceiveAlipayResult(resDict)
        }
    }

    private func didReceiveAlipayResult(_ resDict: [AnyHashable: Any]?) {
        let resultStatus = resDict?["resultStatus"] as? String ?? ""
        let status: String

        switch resultStatus {
        case "9000":
            status = "success"
        case "6001":
            status = "cancelled"
        default:
            status = "failed"
        }

        let response: [String: Any] = [
            "channel": "alipay",
            "status": status,
            "result": resDict ?? [:]
        ]

        pendingResult?(response)
        pendingResult = nil
    }

    private func payWithWechat(orderInfo: String) {
        guard wechatAppId != nil else {
            pendingResult?(FlutterError(code: "NOT_CONFIGURED", message: "Wechat AppId not configured", details: nil))
            pendingResult = nil
            return
        }

        guard WxPayBridge.isWxAppInstalled() else {
            pendingResult?(FlutterError(code: "APP_NOT_INSTALLED", message: "Wechat is not installed", details: nil))
            pendingResult = nil
            return
        }

        guard let orderDict = parseOrderInfo(orderInfo) else {
            pendingResult?(FlutterError(code: "INVALID_ORDER", message: "Invalid order info", details: nil))
            pendingResult = nil
            return
        }

        let partnerId = orderDict["partnerId"] as? String ?? ""
        let prepayId = orderDict["prepayId"] as? String ?? ""
        let package = orderDict["package"] as? String ?? "Sign=WXPay"
        let nonceStr = orderDict["nonceStr"] as? String ?? ""
        let timeStamp = UInt32(orderDict["timeStamp"] as? String ?? "") ?? 0
        let sign = orderDict["sign"] as? String ?? ""

        WxPayBridge.sendPayReq(withPartnerId: partnerId, prepayId: prepayId, package: package, nonceStr: nonceStr, timeStamp: timeStamp, sign: sign)

        pendingResult?(["channel": "wechat", "status": "pending"])
        pendingResult = nil
    }

    private func payWithUnionPay(orderInfo: String) {
        guard let topController = UIApplication.shared.topViewController() else {
            pendingResult?(FlutterError(code: "NO_VIEW_CONTROLLER", message: "No view controller", details: nil))
            pendingResult = nil
            return
        }

        let mode = unionPayTnMode ?? "00"
        unionPayPendingResult = pendingResult
        pendingResult = nil

        let started = UPPaymentControl.default().startPay(
            orderInfo,
            fromScheme: unionPayScheme,
            mode: mode,
            viewController: topController
        )
        if !started {
            let result = unionPayPendingResult
            unionPayPendingResult = nil
            result?(FlutterError(code: "UNIONPAY_ERROR", message: "Failed to start UnionPay", details: nil))
        }
    }

    func handleUnionPayCallback(code: String, data: [AnyHashable: Any]?) {
        guard let result = unionPayPendingResult else { return }
        unionPayPendingResult = nil

        let status: String
        switch code {
        case "success":
            status = "success"
        case "cancel":
            status = "cancelled"
        default:
            status = "failed"
        }

        result(["channel": "unionPay", "status": status])
    }

    func didReceiveWechatPayResponse(errCode: Int, errStr: String) {
        let status: String
        switch errCode {
        case 0:
            status = "success"
        case -2:
            status = "cancelled"
        default:
            status = "failed"
        }

        let response: [String: Any] = [
            "channel": "wechat",
            "status": status,
            "errCode": errCode,
            "errStr": errStr
        ]

        pendingResult?(response)
        pendingResult = nil
    }

    private func handleIsAppInstalled(channel: String) -> Bool {
        switch channel {
        case "alipay":
            return UIApplication.shared.canOpenURL(URL(string: "alipay://")!)
        case "wechat":
            return WxPayBridge.isWxAppInstalled()
        case "unionPay":
            let mode = unionPayTnMode ?? "00"
            return UPPaymentControl.default().isPaymentAppInstalled(mode, withMerchantInfo: "")
        default:
            return false
        }
    }

    private func parseOrderInfo(_ jsonString: String) -> [String: Any]? {
        guard let data = jsonString.data(using: .utf8),
              let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }
        return dict
    }
}

extension MultiPayPlugin: WxPayDelegate {
    public func onWechatPayResponse(withErrCode errCode: Int32, errStr: String) {
        didReceiveWechatPayResponse(errCode: Int(errCode), errStr: errStr)
    }
}

extension UIApplication {
    func topViewController() -> UIViewController? {
        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let window = windowScene.windows.first,
              var topController = window.rootViewController else {
            return nil
        }

        while let presentedController = topController.presentedViewController {
            topController = presentedController
        }

        return topController
    }
}

File: android/app/src/main/kotlin/com/plugin/multipay/MultiPayPlugin.kt
package com.plugin.multipay

import android.app.Activity
import android.content.Context
import android.content.Intent
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleEventObserver
import com.alipay.sdk.app.PayTask
import com.tencent.mm.opensdk.modelpay.PayReq
import com.tencent.mm.opensdk.openapi.IWXAPI
import com.tencent.mm.opensdk.openapi.WXAPIFactory
import com.unionpay.sdk.UPPayActivity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.*

class MultiPayPlugin : FlutterPlugin, MethodChannel.MethodCallHandler, ActivityAware,
    ActivityPluginBinding.OnActivityResultListener {

    private lateinit var context: Context
    private lateinit var channel: MethodChannel
    private var currentActivity: Activity? = null
    private var pendingResult: MethodChannel.Result? = null
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    private var alipayAppId: String? = null
    private var wechatAppId: String? = null
    private var unionPayTnMode: String? = null
    private var isWechatPayPending = false

    companion object {
        private const val UNIONPAY_REQUEST_CODE = 1002
        var wechatAppId: String? = null
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "multi_pay_channel")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "pay" -> handlePay(call, result)
            "isAppInstalled" -> handleIsAppInstalled(call, result)
            "initSdk" -> handleInitSdk(call, result)
            else -> result.notImplemented()
        }
    }

    private fun handlePay(call: MethodCall, result: MethodChannel.Result) {
        val channelName = call.argument<String>("channel")
        val orderInfo = call.argument<String>("orderInfo")

        if (channelName == null || orderInfo == null) {
            result.error("INVALID_ARGS", "Channel and orderInfo are required", null)
            return
        }

        pendingResult = result

        when (channelName) {
            "alipay" -> payWithAlipay(orderInfo)
            "wechat" -> payWithWechat(orderInfo)
            "unionPay" -> payWithUnionPay(orderInfo)
            else -> {
                result.error("UNKNOWN_CHANNEL", "Unknown channel: $channelName", null)
                pendingResult = null
            }
        }
    }

    private fun payWithAlipay(orderInfo: String) {
        val activity = currentActivity
        if (activity == null) {
            pendingResult?.error("NO_ACTIVITY", "Activity is not available", null)
            pendingResult = null
            return
        }

        scope.launch {
            try {
                val alipayTask = PayTask(activity)
                val payResult = alipayTask.payV2(orderInfo, true)

                val resultStatus = payResult["resultStatus"] as? String ?: ""
                val resultCode = when (resultStatus) {
                    "9000" -> "success"
                    "6001" -> "cancelled"
                    else -> "failed"
                }

                val response = hashMapOf<String, Any?>(
                    "channel" to "alipay",
                    "status" to resultCode,
                    "result" to payResult
                )

                pendingResult?.success(response)
            } catch (e: Exception) {
                pendingResult?.error("ALIPAY_ERROR", e.message, null)
            } finally {
                pendingResult = null
            }
        }
    }

    private fun payWithWechat(orderInfo: String) {
        val wechatAppId = this.wechatAppId

        if (wechatAppId == null) {
            pendingResult?.error("NOT_CONFIGURED", "Wechat AppId not configured", null)
            pendingResult = null
            return
        }

        val api: IWXAPI = WXAPIFactory.createWXAPI(context, wechatAppId)

        if (!api.isWXAppInstalled) {
            pendingResult?.error("APP_NOT_INSTALLED", "Wechat is not installed", null)
            pendingResult = null
            return
        }

        try {
            val orderMap = parseOrderInfo(orderInfo)
            if (orderMap.isEmpty()) {
                pendingResult?.error("INVALID_ORDER", "Failed to parse order info", null)
                pendingResult = null
                return
            }

            val req = PayReq().apply {
                appId = orderMap["appId"] as? String ?: wechatAppId
                partnerId = orderMap["partnerId"] as? String
                prepayId = orderMap["prepayId"] as? String
                packageValue = orderMap["package"] as? String ?: "Sign=WXPay"
                nonceStr = orderMap["nonceStr"] as? String
                timeStamp = orderMap["timeStamp"] as? String
                sign = orderMap["sign"] as? String
            }

            api.registerApp(wechatAppId)
            val sent = api.sendReq(req)

            if (!sent) {
                pendingResult?.error("SEND_REQ_FAILED", "Failed to send request to Wechat", null)
                pendingResult = null
            } else {
                isWechatPayPending = true
            }
        } catch (e: Exception) {
            pendingResult?.error("WECHAT_ERROR", e.message, null)
            pendingResult = null
        }
    }

    private fun payWithUnionPay(orderInfo: String) {
        val activity = currentActivity
        if (activity == null) {
            pendingResult?.error("NO_ACTIVITY", "Activity is not available", null)
            pendingResult = null
            return
        }

        try {
            val mode = unionPayTnMode ?: "00"
            val intent = Intent(activity, UPPayActivity::class.java).apply {
                putExtra("merchant", orderInfo)
                putExtra("mode", mode)
            }
            activity.startActivityForResult(intent, UNIONPAY_REQUEST_CODE)
        } catch (e: Exception) {
            pendingResult?.error("UNIONPAY_ERROR", e.message, null)
            pendingResult = null
        }
    }

    private fun handleIsAppInstalled(call: MethodCall, result: MethodChannel.Result) {
        val channelName = call.argument<String>("channel")
        if (channelName == null) {
            result.error("INVALID_ARGS", "Channel is required", null)
            return
        }

        val isInstalled = when (channelName) {
            "alipay" -> isPackageInstalled("com.eg.android.AlipayGphone")
            "wechat" -> isPackageInstalled("com.tencent.mm")
            "unionPay" -> isPackageInstalled("com.unionpay")
            else -> false
        }

        result.success(isInstalled)
    }

    private fun handleInitSdk(call: MethodCall, result: MethodChannel.Result) {
        alipayAppId = call.argument<String>("alipayAppId")
        wechatAppId = call.argument<String>("wechatAppId")
        unionPayTnMode = call.argument<String>("unionPayTnMode")

        wechatAppId?.let { appId ->
            MultiPayPlugin.wechatAppId = appId
            val api = WXAPIFactory.createWXAPI(context, appId)
            api.registerApp(appId)
        }

        result.success(null)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
        when (requestCode) {
            UNIONPAY_REQUEST_CODE -> {
                val status = when (resultCode) {
                    Activity.RESULT_OK -> "success"
                    Activity.RESULT_CANCELED -> "cancelled"
                    else -> "failed"
                }
                pendingResult?.success(mapOf("channel" to "unionPay", "status" to status))
                pendingResult = null
                return true
            }
        }
        return false
    }

    private fun checkWechatResult() {
        if (!isWechatPayPending || pendingResult == null) return

        val result = WechatResultHolder.lastResult
        if (result != null) {
            WechatResultHolder.lastResult = null
            pendingResult?.success(result)
            pendingResult = null
            isWechatPayPending = false
        }
    }

    private fun isPackageInstalled(packageName: String): Boolean {
        return try {
            context.packageManager.getPackageInfo(packageName, 0)
            true
        } catch (e: Exception) {
            false
        }
    }

    private fun parseOrderInfo(jsonString: String): Map<String, Any> {
        return try {
            val jsonObject = org.json.JSONObject(jsonString)
            val map = mutableMapOf<String, Any>()
            jsonObject.keys().forEach { key ->
                map[key] = jsonObject.get(key)
            }
            map
        } catch (e: Exception) {
            emptyMap()
        }
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        currentActivity = binding.activity
        binding.addActivityResultListener(this)
        binding.lifecycle.addObserver(LifecycleEventObserver { _, event ->
            if (event == Lifecycle.Event.ON_RESUME) {
                checkWechatResult()
            }
        })
    }

    override fun onDetachedFromActivityForConfigChanges() {
        currentActivity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        currentActivity = binding.activity
        binding.addActivityResultListener(this)
    }

    override fun onDetachedFromActivity() {
        currentActivity = null
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        scope.cancel()
    }
}